### PR TITLE
Scene improvements (#1651)

### LINKED ISSUE #1651: Scene improvements
The current `sprite_lists` and `name_mapping` members are "public". Some users are modifying these directly causing issues that are hard to debug and understand. We should make them private and add additional tools if needed.

https://github.com/pythonarcade/arcade/blob/98d4caca12768d712cdcf7e04799f6e57ebb9d47/arcade/scene.py#L30-L32

Things like `__len__` and `__delitem__` are candidates here.

On a few occations users have been deleting spritelists from `name_mapping` thinking that this will remove the spritelist. For more dynamic worlds this means you end up with hundres of spritelists in `spritelists` and performance tanks.

### PR DESCRIPTION
**Pull request by gran4**
attempt to fix https://github.com/pythonarcade/arcade/issues/1651

### CODE REVIEW

**Comment by gran4** [comment]
I think the warnings I added are breaking the tests, it says "Warning, treated as error:".

File: arcade/scene.py [hunk 1]
```diff
@@ -20,17 +21,18 @@ class Scene:
...
-        :sprite_lists: A list of `SpriteList` objects. The order of this list is the
+        :DON'T CHANGE __sprite_lists or __name_mapping
+        :__sprite_lists: A list of `SpriteList` objects. The order of this list is the
                        order in which they will be drawn.
-        :name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
-                       as the `sprite_lists` attribute, but is a mapping of them by name. This is
-                       not necessarily in the same order as the `sprite_lists` attribute.
+        :__name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
+                       as the `__sprite_lists` attribute, but is a mapping of them by name. This is
+                       not necessarily in the same order as the `__sprite_lists` attribute.(Dictionaries have no order)
```
Comment (einarf) [inline arcade/scene.py]: If these are non-public this entire section should just be removed.

File: arcade/scene.py [hunk 2]
```diff
@@ -20,17 +21,18 @@ class Scene:
...
     """
 
     def __init__(self) -> None:
-        self.sprite_lists: List[SpriteList] = []
-        self.name_mapping: Dict[str, SpriteList] = {}
-
+        self.__sprite_lists: List[SpriteList] = []
+        self.__name_mapping: Dict[str, SpriteList] = {}
```
Comment (einarf) [inline arcade/scene.py]: We use single underscore for "private" members.

File: arcade/scene.py (hunk 1 above)
Comment (gran4) [reply arcade/scene.py]: that's true

File: arcade/scene.py (hunk 2 above)
Comment (gran4) [reply arcade/scene.py]: Fixed

File: arcade/scene.py (hunk 1 above)
Comment (gran4) [reply arcade/scene.py]: Fixed. @einarf.
Thanks

File: arcade/scene.py [hunk 3]
```diff
@@ -20,17 +21,13 @@ class Scene:
     https://api.arcade.academy/en/latest/tutorials/views/index.html
 
     Attributes:
-        :sprite_lists: A list of `SpriteList` objects. The order of this list is the
-                       order in which they will be drawn.
-        :name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
-                       as the `sprite_lists` attribute, but is a mapping of them by name. This is
-                       not necessarily in the same order as the `sprite_lists` attribute.
+        :DON'T CHANGE _sprite_lists or _name_mapping
```
Comment (einarf) [inline arcade/scene.py]: Just remove this whole attribute section. It's already implied that user should not access these members unless they know exactly what they are doing.

File: arcade/scene.py [hunk 4]
```diff
@@ -113,8 +110,11 @@ def add_sprite_list(
         """
         if sprite_list is None:
             sprite_list = SpriteList(use_spatial_hash=use_spatial_hash)
-        self.name_mapping[name] = sprite_list
-        self.sprite_lists.append(sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
Comment (einarf) [inline arcade/scene.py:113]: It's usually better to have a hard error than doing a warning. I'm not sure if this should be a warning/error. Might need some more input from user users/devs. @Cleptomania have any opinion on this?

File: arcade/scene.py [hunk 5]
```diff
@@ -138,10 +138,13 @@ def add_sprite_list_before(
         """
         if sprite_list is None:
             sprite_list = SpriteList(use_spatial_hash=use_spatial_hash)
-        self.name_mapping[name] = sprite_list
-        before_list = self.name_mapping[before]
-        index = self.sprite_lists.index(before_list)
-        self.sprite_lists.insert(index, sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
Comment (einarf) [inline arcade/scene.py:141]: Again: We need to figure out if this should be allowed or not.

File: arcade/scene.py [hunk 6]
```diff
@@ -190,10 +193,13 @@ def add_sprite_list_after(
         """
         if sprite_list is None:
             sprite_list = SpriteList(use_spatial_hash=use_spatial_hash)
-        self.name_mapping[name] = sprite_list
-        after_list = self.name_mapping[after]
-        index = self.sprite_lists.index(after_list) + 1
-        self.sprite_lists.insert(index, sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
Comment (einarf) [inline arcade/scene.py:196]: Again: We need to figure out if this should be allowed or not.

File: arcade/scene.py [hunk 7]
```diff
@@ -307,7 +313,7 @@ def draw(self, names: Optional[List[str]] = None, **kwargs) -> None:
         If `names` parameter is provided then only the specified SpriteLists
         will be drawn. They will be drawn in the order that the names in the
         list were arranged. If `names` is not provided, then every SpriteList
-        in the scene will be drawn according the order of the main sprite_lists
+        in the scene will be drawn according the order of the main _sprite_lists
```
Comment (einarf) [inline arcade/scene.py:314]: Should probably be phrased wiout mentioning the internal member.

**Comment by gran4** [comment]
oops. I meant to wait. Is there a way to pull it back?

**Comment by einarf** [comment]
You can just keep pushing changes. We will squash everything into one commit when merging, so it doesn't matter if it's messy commit history in the PR.

**Comment by gran4** [comment]
I meant to fix another issue(https://github.com/pythonarcade/arcade/issues/1410)

File: arcade/scene.py (hunk 4 above)
Comment (gran4) [reply arcade/scene.py:113]: There might be a case where the user doesn't care.

File: arcade/scene.py (hunk 3 above)
Comment (gran4) [reply arcade/scene.py]: Fixed

**Comment by einarf** [comment]
> I meant to fix another issue(#1410)

`git revert 231541f287c25e209f96ad015289ecaee92019d0` maybe?

**Comment by gran4** [comment]
I just added another commit that reverts my changes.

File: arcade/scene.py (hunk 7 above)
Comment (gran4) [reply arcade/scene.py:314]: Change "the order of the main sprite_lists" to "the order of the sprite lists". Or just delete that? Or something else?

**Pull request closed without merging**